azurelinuxagent/common/protocol/wire.py

```diff
@@ -908,49 +918,59 @@ def _call_hostplugin_with_container_check(self, host_func):
+        """
+        Calls the host_func on host channel with retries for stale goal state and handles any exceptions, consistent with the caller for direct channel.
+        """
+        ret = None
+        try:
+            ret = self._call_hostplugin_with_container_check(host_func)
+        except Exception as error:
+            logger.periodic_info(logger.EVERY_HOUR, "[PERIODIC] Request failed using the host channel. Error: {0}".format(ustr(error)))
+
         return ret
 
-    def send_request_using_appropriate_channel(self, direct_func, host_func):
-        # A wrapper method for all function calls that send HTTP requests. The purpose of the method is to
-        # define which channel to use, direct or through the host plugin. For the host plugin channel,
-        # also implement a retry mechanism.
+    @staticmethod
+    def __send_request_using_direct_channel(direct_func):
+        """
+        Calls the direct_func on direct channel and handles any exceptions, consistent with the caller for host channel.
+        """
+        ret = None
+        try:
+            ret = direct_func()
 
-        # By default, the direct channel is the default channel. If that is the case, try getting a response
-        # through that channel. On failure, fall back to the host plugin channel.
+            if ret is None or not ret:
+                logger.periodic_info(logger.EVERY_HOUR, "[PERIODIC] Request failed using the direct channel.")
+                return None
+        except Exception as error:
+            logger.periodic_info(logger.EVERY_HOUR, "[PERIODIC] Request failed using the direct channel. Error: {0}".format(ustr(error)))
 
-        # When using the host plugin channel, regardless if it's set as default or not, try sending the request first.
-        # On specific failures that indicate a stale goal state (such as resource gone or invalid container parameter),
-        # refresh the goal state and try again. If successful, set the host plugin channel as default. If failed,
-        # raise the exception.
+        return ret
 
-        # NOTE: direct_func and host_func are passed as lambdas. Be careful about capturing goal state data in them as
-        # they will not be refreshed even if a goal state refresh is called before retrying the host_func.
+    def send_request_using_appropriate_channel(self, direct_func, host_func):
+        """
+        Determines which communication channel to use. By default, the primary channel is direct, host channel is secondary.
+        We call the primary channel first and return on success. If primary fails, we try secondary. If secondary fails,
+        we return and *don't* switch the default channel. If secondary succeeds, we change the default channel.
+        This method doesn't raise since the calls to direct_func and host_func are already wrapped and handle any exceptions.
+        Possible return values are manifest, artifacts profile, boolean or None.
```

Comment:
Correct me if I'm wrong, but this function will never return False right? 
If that understanding is correct, could you update the docstring to be more specific - 

>Possible return values are manifest, artifacts profile, **True** or None.

Reply:
It can be False, `direct_func` or `host_func` can be calling `azurelinuxagent.common.protocol.wire.WireClient.stream` which returns False.

Reply:
True, but for both functions (`__send_request_using_direct_channel` and `__send_request_using_host_channel`) you have this check which ensures it can never be False right? (Not sure if I'm missing something here)  

```python
            if ret is None or not ret:
                logger.periodic_info(logger.EVERY_HOUR, "[PERIODIC] Request failed using the direct channel.")
                return None

            if ret is None or not ret:
                raise Exception("Request failed using the host channel.")
```